src/error.rs

```diff
@@ -0,0 +1,88 @@
+//! Web3 Error
+
+#![allow(unknown_lints)]
+#![allow(missing_docs)]
+
+use std::io;
+use serde_json;
+use rpc;
+
+error_chain! {
+  foreign_links {
+    Io(io::Error);
+  }
+  errors {
+    Unreachable {
+      description("server is unreachable"),
+      display("Server is unreachable"),
+    }
+    Decoder(e: String) {
+      description("decoder error"),
+      display("Decoder error: {}", e),
+    }
+    InvalidResponse(e: String) {
+      description("invalid response"),
+      display("Got invalid response: {}", e),
+    }
+    Transport(e: String) {
+      description("transport error"),
+      display("Transport error: {}", e),
+    }
+    // TODO [ToDr] Move to foreign_links
+    Rpc(e: rpc::Error) {
+      description("rpc error"),
+      display("RPC error: {:?}", e),
+    }
+    Internal {
+      description("web3 internal error"),
+      display("Internal Web3 error"),
+    }
+  }
+}
+
+impl From<serde_json::Error> for Error {
+  fn from(err: serde_json::Error) -> Self {
+    ErrorKind::Decoder(format!("{:?}", err)).into()
+  }
+}
+
+impl Clone for Error {
+  fn clone(&self) -> Self {
```

Comment:
why is it needed?

Reply:
It's needed for batch requests, where you can do:
`let accounts = web3.eth().accounts();`
`let batch = web3.transport().send_batch()`;
and both futures need to resolve to the same value.